### PR TITLE
add README.md

### PR DESCRIPTION
**Pull request by jcaip**
add README.md to sparsity folder

### CODE REVIEW

**Comment by jcaip** [comment]
adding images here for hosting lol:
Screenshot 2024-04-22 at 1 46 15 PM
Screenshot 2024-04-22 at 1 44 13 PM
Screenshot 2024-04-22 at 1 44 03 PM

File: torchao/sparsity/README.md [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+  </tr>
+</table>
+
+
+_Table 4.4: Description of some common sparsity patterns. _
+
+
+# Design
+
+Pruning, like quantization, is an accuracy/performance trade-off, where we care not only about the speedup but also on the accuracy degradation of our architecture optimization technique.
```
Comment (jerryzh168) [inline torchao/sparsity/README.md:702]: should we rename the folder to pruning

Comment (jcaip) [reply torchao/sparsity/README.md:676]: I think sparsity is the more widely used term, so let's keep it called that. But I'll change Pruning -> Sparsity where it makes sense in the README

File: torchao/sparsity/README.md [hunk 2]
```diff
@@ -0,0 +1,664 @@
+# torchao sparsity
+
+Sparsity is the technique of removing parameters from a neural network in order to reduce its memory overhead or latency. By carefully choosing the elements that are removed, one can achieve significant reduction in memory overhead and latency, while paying a reasonably low or no price in terms of model quality (accuracy / f1).
```
Comment (cpuhrsch) [inline torchao/sparsity/README.md:3]: I'd call pruning the "technique of removing parameters from a neural network in order to reduce its memory overhead or latency"

File: torchao/sparsity/README.md [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+
+1. Train sparse models from scratch with hardware acceleration, with minimal accuracy loss.
+2. Recover accuracy loss of pruned model with custom pruning algorthim.
+3. Accelerate masked/pruned models on sparsity-supported hardware to realize performance improvements.
+
+# Design
+
+Sparsity, like quantization, is an accuracy/performance trade-off, where we care not only about the speedup but also on the accuracy degradation of our architecture optimization technique.
+
+In quantization, the theoretical performance gain is generally determined by the data type that we are quantizing to - quantizing from float32 to float16 yields a theoretical 2x speedup. For pruning/sparsity, the analogous variable would be the sparsity level/ sparsity pattern. For semi-structured, the sparsity level is fixed at 50%, so we expect a theoretical 2x improvement. For block-sparse matrices and unstructured sparsity, the speedup is variable and depends on the sparsity level of the tensor.
```
Comment (cpuhrsch) [inline torchao/sparsity/README.md:25]: nit: It's roughly a theoretical speedup of 2x. Or put differently, 2x is a very basic estimate just because of the reduce amount of memory that needs to processed. In practice it can vary quite a bit. It could even be a lot more, because it allows you to use faster caches, etc.

File: torchao/sparsity/README.md [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+One key difference between sparsity and quantization is in how the accuracy degradation is determined: The accuracy degradation of quantization is determined by the scale and zero_point chosen. However, in pruning the accuracy degradation is determined by the mask. By carefully choosing the specified elements and retraining the network, pruning can achieve negligible accuracy degradation and in some cases even provide a slight accuracy gain. This is an active area of research with no agreed-upon consensus. We expect users will have a target sparsity pattern and mind and to prune to that pattern.
```
Comment (cpuhrsch) [inline torchao/sparsity/README.md:27]: This is a bit biased towards affine quantization and sparsity aware training specifically for matrix multiplication. There's many other variables that influences accuracy degradation. For example the operation used and the distribution of input values.

The measure that is model quality is the same between sparsity and quantization. Some of the mitigation techniques are the same too (e.g. quantization or sparsity aware training). Where it differs, I'd say, is that sparsity explicitly relies on approximating a sum of numbers (hence the focus on zero), whereas in quantization you avoid allocating bits for unused numerical ranges/unnecessary numerical fidelty.

File: torchao/sparsity/README.md [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+
+Given a target sparsity pattern, pruning a model can then be thought of as two separate subproblems:
+
+* How can I find a set of sparse weights which satisfy my target sparsity pattern that minimize the accuracy degradation of my model?
```
Comment (cpuhrsch) [inline torchao/sparsity/README.md:31]: Right, so this first part is what I'd call pruning.

File: torchao/sparsity/README.md [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+* How can I accelerate my sparse weights for inference and reduced memory overhead?
```
Comment (cpuhrsch) [inline torchao/sparsity/README.md:32]: And then sparsity can be the task of accelerating pruned weights. It's not always necessary to use a sparse layout or sparse kernel. Sometimes you can prune in ways that obviate these specialized techniques. For example, you can just skip an entire layer.

File: torchao/sparsity/README.md [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+
+Fundamentally, the flow works by manipulating `torch.Tensors`. In the frontend, we specify the tensors by their fully-qualified-name in a sparse_config dictionary. The frontend is designed to follow the quantization API, with a `prepare` function, which attaches FakeSparsity paramerizations to the tensors specified in the config.
+
+FakeSparsity is a parameterization which simulates unstructured sparsity, where each element has a mask. Because of this, we can use it to simulate any sparsity pattern we want.
+
+The user will then train the prepared model using their own custom code, calling .step() to update the mask if necessary. Once they’ve found a suitable mask, they call `squash_mask()` to fuse the mask into the weights, creating a dense tensor with 0s in the right spot.
```
Comment (msaroufim) [inline torchao/sparsity/README.md:51]: Not sure I follow this line, what's step? Also this seems to indicate that people need to change their training code and if so how?

is this line also necessary for people only interested in accelerated inference

File: torchao/sparsity/README.md [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+
+Users will then convert their model for accelerated sparse inference by either using the quantization flow for quantized block sparse CPU inference or by calling `to_sparse_semi_structured` on the specified weight tensors.
```
Comment (msaroufim) [inline torchao/sparsity/README.md:53]: worth mentioning GPU acceleration as well?

**Review by msaroufim — COMMENTED** [review]
Really enjoyed reading this, it is missing code samples but I believe what you intended to write was closer to a survey of sparsity and the parameter space a library should be in which case I believe this does the job well

File: torchao/sparsity/README.md (hunk 4 above)
Comment (jcaip) [reply torchao/sparsity/README.md:27]: I'll add some more context to the end of this section, but for this and the comment above, I want to keep this as newbie friendly as possible, so I think it's okay to have a relative flawed / forceful analogy to make a point. 

I think explaining things in the most faithful way introduces a lot of jargon, which is kind of overwhelming.

File: torchao/sparsity/README.md (hunk 7 above)
Comment (jcaip) [reply torchao/sparsity/README.md]: I updated with a code sample, that should make this a bit easier to follow.

**Comment by msaroufim** [comment]
So we have a docs page now https://github.com/pytorch/ao/blob/jcaip/sparsity-readme/docs/source/sparsity.rst
pytorch.org/ao

**Comment by jcaip** [comment]
> So we have a docs page now https://github.com/pytorch/ao/blob/jcaip/sparsity-readme/docs/source/sparsity.rst pytorch.org/ao

I think this is conceptually the right long term home for most of the stuff in the README, but I feel like this information will get lost right now vs if we put it in the README.

**Review by msaroufim — APPROVED** [review]
